File: tests/test_integrations/test_langchain/apps/langchain_retriever_app.py
"""
Retriever LangChain App: RAG with deterministic retriever
Complexity: MEDIUM - Tests retriever spans with deterministic results

Uses a fake retriever that returns fixed documents.
"""

from langchain_core.language_models.fake_chat_models import (
    FakeMessagesListChatModel,
)
from langchain_core.messages import AIMessage, HumanMessage
from langchain_core.documents import Document
from langchain_core.retrievers import BaseRetriever
from langchain_core.runnables import RunnableConfig, RunnableLambda
from langchain_core.callbacks.manager import CallbackManagerForRetrieverRun
from typing import List


class DeterministicRetriever(BaseRetriever):
    """A retriever that returns fixed documents based on query keywords."""

    documents: dict = {
        "python": [
            Document(
                page_content="Python is a high-level programming language known for its simplicity.",
                metadata={"source": "doc1"},
            ),
            Document(
                page_content="Python supports multiple programming paradigms including procedural and OOP.",
                metadata={"source": "doc2"},
            ),
        ],
        "langchain": [
            Document(
                page_content="LangChain is a framework for developing applications powered by language models.",
                metadata={"source": "doc3"},
            ),
            Document(
                page_content="LangChain provides tools for chaining LLM calls and integrating with external data.",
                metadata={"source": "doc4"},
            ),
        ],
        "default": [
            Document(
                page_content="This is a general document about AI and machine learning.",
                metadata={"source": "doc5"},
            ),
            Document(
                page_content="Machine learning enables computers to learn from data without explicit programming.",
                metadata={"source": "doc6"},
            ),
        ],
    }

    def _get_relevant_documents(
        self, query: str, *, run_manager: CallbackManagerForRetrieverRun
    ) -> List[Document]:
        """Get documents based on query keywords."""
        query_lower = query.lower()

        if "python" in query_lower:
            return self.documents["python"]
        elif "langchain" in query_lower:
            return self.documents["langchain"]
        else:
            return self.documents["default"]


def _create_rag_chain(response_text: str):
    """Create a RAG chain with the specified response."""
    retriever = DeterministicRetriever()

    def run_rag(inputs: dict, config: RunnableConfig = None):
        messages = inputs.get("messages", [])

        # Extract query from messages
        query = ""
        for msg in reversed(messages):
            if isinstance(msg, HumanMessage):
                query = msg.content
                break
            elif isinstance(msg, tuple) and msg[0] == "human":
                query = msg[1]
                break

        # Retrieve documents
        docs = retriever.invoke(query, config=config)

        # Format context
        context = "\n\n".join([doc.page_content for doc in docs])

        # Create LLM with appropriate response based on query
        if "python" in query.lower():
            final_response = "Based on the retrieved documents, Python is a high-level programming language known for its simplicity and support for multiple paradigms."
        elif "langchain" in query.lower():
            final_response = "According to the documents, LangChain is a framework for building LLM-powered applications with tools for chaining calls and data integration."
        else:
            final_response = "Based on the context, AI and machine learning enable computers to learn patterns from data."

        llm = FakeMessagesListChatModel(
            responses=[AIMessage(content=final_response)]
        )

        # Create augmented prompt
        augmented_messages = list(messages) + [
            HumanMessage(
                content=f"Context:\n{context}\n\nAnswer based on the context above."
            )
        ]

        # Generate response
        response = llm.invoke(augmented_messages, config=config)

        return {
            "messages": list(messages) + [response],
            "context": context,
            "source_documents": docs,
        }

    return RunnableLambda(run_rag)


chain = _create_rag_chain("default")


def invoke_rag_app(inputs: dict, config: RunnableConfig = None):
    """Invoke the RAG app."""
    return chain.invoke(inputs, config=config)


async def ainvoke_rag_app(inputs: dict, config: RunnableConfig = None):
    """Async invoke the RAG app."""
    return await chain.ainvoke(inputs, config=config)

File: typescript/test/test-integrations/test-langchain/apps/langchain-retriever-app.ts
import { BaseRetriever, BaseRetrieverInput } from "@langchain/core/retrievers";
import { Document } from "@langchain/core/documents";
import { ChatOpenAI } from "@langchain/openai";
import {
  HumanMessage,
  SystemMessage,
  BaseMessage,
} from "@langchain/core/messages";
import { RunnableLambda, RunnableConfig } from "@langchain/core/runnables";
import { CallbackManagerForRetrieverRun } from "@langchain/core/callbacks/manager";

class DeterministicRetriever extends BaseRetriever {
  lc_namespace = ["langchain", "retrievers"];
  documents: Record<string, Document[]> = {
    python: [
      new Document({
        pageContent:
          "Python is a high-level programming language known for its simplicity.",
        metadata: { source: "doc1" },
      }),
      new Document({
        pageContent:
          "Python supports multiple programming paradigms including procedural and OOP.",
        metadata: { source: "doc2" },
      }),
    ],
    langchain: [
      new Document({
        pageContent:
          "LangChain is a framework for developing applications powered by language models.",
        metadata: { source: "doc3" },
      }),
      new Document({
        pageContent:
          "LangChain provides tools for chaining LLM calls and integrating with external data.",
        metadata: { source: "doc4" },
      }),
    ],
    default: [
      new Document({
        pageContent:
          "This is a general document about AI and machine learning.",
        metadata: { source: "doc5" },
      }),
      new Document({
        pageContent:
          "Machine learning enables computers to learn from data without explicit programming.",
        metadata: { source: "doc6" },
      }),
    ],
  };

  constructor(
    fields?: BaseRetrieverInput & { metadata?: Record<string, any> },
  ) {
    super(fields);
    if (fields?.metadata) this.metadata = fields.metadata;
  }

  async _getRelevantDocuments(
    query: string,
    runManager?: CallbackManagerForRetrieverRun,
  ): Promise<Document[]> {
    const queryLower = query.toLowerCase();
    if (queryLower.includes("python")) return this.documents.python;
    if (queryLower.includes("langchain")) return this.documents.langchain;
    return this.documents.default;
  }
}

const retriever = new DeterministicRetriever();
const retrieverWithMetricCollection = new DeterministicRetriever({
  metadata: { metric_collection: "retriever_quality" },
});
const llm = new ChatOpenAI({ modelName: "gpt-4o-mini", temperature: 0 });

const runRagChain = async (
  inputs: { messages: BaseMessage[] },
  currentRetriever: any,
  config?: RunnableConfig,
) => {
  const messages = inputs.messages || [];
  const query =
    messages.length > 0 ? messages[messages.length - 1].content.toString() : "";
  const docs = await (currentRetriever.invoke as any)(query, config);
  const context = docs.map((d: any) => d.pageContent).join("\n\n");

  const augmentedMessages = [
    {
      role: "system",
      content:
        "You are a helpful assistant. Answer the user's question based ONLY on the provided context. Be concise and factual.",
    },
    ...messages.slice(0, -1),
    {
      role: "user",
      content: `Context:\n${context}\n\nAnswer based on the context above.\nQuery: ${query}`,
    },
  ];

  const response = await (llm.invoke as any)(augmentedMessages, config);
  return { messages: [...messages, response], context, source_documents: docs };
};

const ragChain = new RunnableLambda({
  func: (inputs: any, config?: RunnableConfig) =>
    runRagChain(inputs, retriever, config),
}).withConfig({ runName: "rag_chain" });
const ragChainWithMetricCollection = new RunnableLambda({
  func: (inputs: any, config?: RunnableConfig) =>
    runRagChain(inputs, retrieverWithMetricCollection, config),
}).withConfig({ runName: "rag_chain" });

export const invokeRagApp = async (
  inputs: { messages: BaseMessage[] },
  config?: RunnableConfig,
) => await ragChain.invoke(inputs, config);
export const invokeRagAppWithMetricCollection = async (
  inputs: { messages: BaseMessage[] },
  config?: RunnableConfig,
) => await ragChainWithMetricCollection.invoke(inputs, config);
